File: simple_ds/max_stack.cpp
#include <iostream>
#include <stack>

using namespace std;

int main() {
    stack<int> regular, max;
    int n, op;
    string input;
    string POP("pop");
    string MAX("max");
    string PUSH("push");
    cin >> n;
    for (int i = 0; i < n; i++) {
        cin >> input;
        if (input == PUSH) {
            cin >> op;
            regular.push(op);
            if (max.empty()) {
                max.push(op);
            } else {
                if (max.top() >= op) {
                    max.push(max.top());
                } else {
                    max.push(op);
                }
            }
        }
        else if (input == MAX) {
            cout << max.top() << endl;
        }
        else if (input == POP) {
            regular.pop();
            max.pop();
        }
    }
    return 0;
}

File: leetcode/src/main/java/edu/agus/leetcode/easy/MaxStack.java
package edu.agus.leetcode.easy;

import java.util.ArrayDeque;
import java.util.Deque;
import java.util.Stack;

public class MaxStack {
  Deque<Integer> stack = new ArrayDeque<>();
  Deque<Integer> max = new ArrayDeque<>();

  public MaxStack() {
  }

  public void push(int x) {
    stack.push(x);
    if (max.isEmpty() || max.peek() <= x) {
      max.push(x);
    } else {
      max.push(max.peek());
    }
  }

  public int pop() {
    max.pop();
    return stack.pop();
  }

  public int top() {
    return stack.peek();
  }

  public int peekMax() {
    return max.peek();
  }

  public int popMax() {
    int max = peekMax();
    Deque<Integer> buffer = new ArrayDeque<>();
    while (top() != max) buffer.push(pop());
    pop();
    while (!buffer.isEmpty()) push(buffer.pop());
    return max;
  }
}
